### PR TITLE
Add learning rate decay

### PR DESCRIPTION
**Pull request by jacquesqiao**
project: https://github.com/PaddlePaddle/Paddle/issues/7769

### CODE REVIEW

File: python/paddle/v2/fluid/optimizer.py [hunk 1]
```diff
@@ -33,16 +33,44 @@ class Optimizer(object):
...
         # Dictionary of accumulators. Some optimizer subclasses need to
         # allocate and manage extra variables associated with the parameters
         # to train. These variables are called accumulators.
         # {accum_name : { paramter_name : accumulator_for_parameter, ...}, ...}
         self._accumulators = defaultdict(lambda: dict())
         self.helper = None
 
+    def _create_global_learning_rate(self):
+        if isinstance(self._learning_rate, float):
+            self._global_lr_var = self.helper.create_global_variable(
```
Comment (QiJune) [inline python/paddle/v2/fluid/optimizer.py]: `create_global_var` combines `helper.create_global_variable` and `helper.set_variable_initializer`. So line51-line59 can be optimized.

File: python/paddle/v2/fluid/learning_rate_decay.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    lr = learning_rate
+    if isinstance(lr, float):
+        lr = layers.create_global_var(
+            shape=[1], value=learning_rate, dtype='float32')
+    elif isinstance(lr, Variable):
+        lr = learning_rate
+    else:
+        raise ValueError("unsupported learning rate type : %s",
+                         type(learning_rate))
```
Comment (reyoung) [inline python/paddle/v2/fluid/learning_rate_decay.py]:
```python
if isinstance(lreaning_rate, float):
   learning_rate = layers.create_global_var(shape=[1], value=learning_rate, dtype='float32')

if not isinstance(learning_rate, Variable):
   raise TypeError("...")

```

File: python/paddle/v2/fluid/learning_rate_decay.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    decay_steps_var = layers.fill_constant(
+        shape=[1], dtype='float32', value=float(decay_steps))
+    decay_rate_var = layers.fill_constant(
+        shape=[1], dtype='float32', value=float(decay_rate))
+    div_res = global_step / decay_steps_var
+    if staircase:
+        div_res = layers.floor(x=div_res)
+    const_one = layers.fill_constant(shape=[1], dtype='float32', value=float(1))
+
+    return lr / (const_one + decay_rate_var * div_res)
```
Comment (reyoung) [inline python/paddle/v2/fluid/learning_rate_decay.py]:
```python
div_res = global_step / decay_steps
if starcase:
  div_res = layer.floor(x=div_res)
return lr / (1 + decay_rate * div_res)
```

File: python/paddle/v2/fluid/optimizer.py [hunk 4]
```diff
@@ -54,14 +82,20 @@ def _create_param_lr(self, param_and_grad):
...
         param_lr_shape = [1]
         param_lr_var = self.helper.create_global_variable(
-            name=unique_name("learning_rate"),
+            name=unique_name("learning_rate_" + param.name),
             dtype='float32',
             shape=param_lr_shape,
             lod_level=1,
             persistable=True)
-        param_lr = param_lr * self._learning_rate
+        param.block.program.global_block().append_op(
```
Comment (QiJune) [inline python/paddle/v2/fluid/optimizer.py]: A elementwise-mul operator?
```
param_lr_var = param_lr  * self._global_lr_var
```

File: python/paddle/v2/fluid/optimizer.py [hunk 5]
```diff
@@ -54,14 +82,20 @@ def _create_param_lr(self, param_and_grad):
...
+            type="scale",
+            inputs={"X": self._global_lr_var},
+            outputs={"Out": param_lr_var},
+            attrs={"scale": param_lr})
+        # give param learning rate a default value 0.0
```
Comment (QiJune) [inline python/paddle/v2/fluid/optimizer.py]: What's the purpose of default value 0.0?

File: python/paddle/v2/fluid/optimizer.py [hunk 6]
```diff
@@ -33,16 +33,44 @@ class Optimizer(object):
...
+                var=self._global_lr_var,
+                initializer=Constant(self._learning_rate))
+        elif isinstance(self._learning_rate, framework.Variable):
+            self._global_lr_var = self._learning_rate
+        else:
+            raise ValueError("learning rate should be a float or a Variable, "
+                             "actual type is %s", type(self._global_lr_var))
+
+    @property
+    def learning_rate(self):
```
Comment (reyoung) [inline python/paddle/v2/fluid/optimizer.py]: Maybe `global_learning_rate()` is more accuracy.

File: python/paddle/v2/fluid/learning_rate_decay.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+import layers
+from framework import Variable
+
+__all__ = ['exponential_decay', ]
+"""
+When training a model, it's often useful to decay the
+learning rate during training process, this is called
+learning_rate_decay. There are many strategies to do
+this, this module will provide some classic method.
```
Comment (abhinavarora) [inline python/paddle/v2/fluid/learning_rate_decay.py]: classic -> `classical`

File: python/paddle/v2/fluid/learning_rate_decay.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+User can also implement their own learning_rate_decay
+strategy according to this module.
+"""
+
+
+def exponential_decay(learning_rate,
```
Comment (abhinavarora) [inline python/paddle/v2/fluid/learning_rate_decay.py]: I think we should implement a base class because both `exponential_decay` and `inverse_time_decay` have a lot of code in common. We can create a base class that contains most of the boilerplate code. The inherited classes, `exponential_decay` , `inverse_time_decay` and any other future learning rate decay schemes can only implement the mathematics for the decay/ What are your thoughts on this?

File: python/paddle/v2/fluid/optimizer.py [hunk 9]
```diff
@@ -33,16 +33,44 @@ class Optimizer(object):
     but need to use one of it's implementation.
     """
 
-    def __init__(self, global_step=None, regularization=None):
+    def __init__(self, learning_rate, global_step=None, regularization=None):
+        assert learning_rate is not None
         self._global_step = global_step
         self.regularization = regularization
+        self._learning_rate = learning_rate
```
Comment (abhinavarora) [inline python/paddle/v2/fluid/optimizer.py]: It would be good to have some comments here.

File: python/paddle/v2/fluid/tests/book/test_label_semantic_roles.py [hunk 10]
```diff
@@ -155,9 +155,18 @@ def main():
             name='crfw', learning_rate=mix_hidden_lr))
     avg_cost = fluid.layers.mean(x=crf_cost)
 
+    global_step = fluid.layers.create_global_var(
+        shape=[1], value=0, dtype='float32')
     # TODO(qiao)
     # check other optimizers and check why out will be NAN
-    sgd_optimizer = fluid.optimizer.SGD(learning_rate=0.0001)
+    sgd_optimizer = fluid.optimizer.SGD(
+        learning_rate=fluid.learning_rate_decay.exponential_decay(
```
Comment (abhinavarora) [inline python/paddle/v2/fluid/tests/book/test_label_semantic_roles.py]: Will the old syntax also work? For example, will this line work with the new code `sgd_optimizer = fluid.optimizer.SGD(learning_rate=0.0001)`

Comment (jacquesqiao) [reply python/paddle/v2/fluid/tests/book/test_label_semantic_roles.py]: yes, it will work, _create_global_learning_rate will handle it if learning_rate is a float value

File: python/paddle/v2/fluid/learning_rate_decay.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import layers
+from framework import Variable
+
+__all__ = ['exponential_decay', ]
```
[the remark this answers is not in the document]
Comment (jacquesqiao) [reply python/paddle/v2/fluid/learning_rate_decay.py]: done

File: python/paddle/v2/fluid/learning_rate_decay.py (hunk 8 above)
Comment (jacquesqiao) [reply python/paddle/v2/fluid/learning_rate_decay.py]: the common code has been removed since we overload some Python Operator such as "+" "/" "*", it will convert the value in Python to Variable implicitly. So that codes are all removed

File: python/paddle/v2/fluid/optimizer.py (hunk 1 above)
Comment (jacquesqiao) [reply python/paddle/v2/fluid/optimizer.py]: done

File: python/paddle/v2/fluid/optimizer.py (hunk 4 above)
Comment (jacquesqiao) [reply python/paddle/v2/fluid/optimizer.py]: done

File: python/paddle/v2/fluid/optimizer.py (hunk 6 above)
Comment (jacquesqiao) [reply python/paddle/v2/fluid/optimizer.py]: done

File: python/paddle/v2/fluid/optimizer.py (hunk 5 above)
Comment (jacquesqiao) [reply python/paddle/v2/fluid/optimizer.py]: removed

File: python/paddle/v2/fluid/learning_rate_decay.py (hunk 3 above)
Comment (jacquesqiao) [reply python/paddle/v2/fluid/learning_rate_decay.py]: done

File: python/paddle/v2/fluid/learning_rate_decay.py (hunk 2 above)
Comment (jacquesqiao) [reply python/paddle/v2/fluid/learning_rate_decay.py]: done
